Assignments/Midterm.py

```diff
@@ -0,0 +1,687 @@
+            ]
+
+    return df
+
+
+def brute_force(pred1, pred2, resp, data):
+    # df = pd.DataFrame(
+    #     columns=[
+    #         "x_1",
+    #         "x_2",
+    #         "Bin_Pop",
+    #         "Bin_Mean",
+    #         "Pop_Mean",
+    #         "Pop_Proportion",
+    #         "Mean_Sq_Diff",
+    #         "Mean_Sq_Diff_Weighted",
+    #     ]
+    # )
+    copy_data = data.copy()
+    if is_continuous(data, pred1):
+        copy_data["bins1"] = pd.cut(copy_data[pred1], 10, labels=False)
+        if is_continuous(data, pred2):
+            copy_data["bins2"] = pd.cut(copy_data[pred2], 10, labels=False)
+            calc_matrix = np.zeros((11, 11))
+            for x_1, val1 in enumerate(copy_data["bins1"].unique()):
+                for x_2, val2 in enumerate(copy_data["bins2"].unique()):
+                    bin_mean = copy_data[resp][
+                        (copy_data["bins1"] == x_1) & (copy_data["bins2"] == x_2)
+                    ].mean()
+                    calc_matrix[x_1, x_2] = bin_mean
+        else:
+            calc_matrix = np.zeros((11, len(copy_data[pred2].unique())))
+            for x_1, val1 in enumerate(copy_data["bins1"].unique()):
+                for x_2, val2 in enumerate(copy_data[pred2].unique()):
+                    bin_mean = copy_data[resp][
+                        (copy_data["bins1"] == x_1) & (copy_data[pred2] == val2)
+                    ].mean()
+                    calc_matrix[x_1, x_2] = bin_mean
+    else:
+        if is_continuous(data, pred2):
+            calc_matrix = np.zeros((len(copy_data[pred1].unique()), 11))
+            copy_data["bins2"] = pd.cut(copy_data[pred2], 10, labels=False)
+            for x_1, val1 in enumerate(copy_data[pred1].unique()):
+                for x_2, val2 in enumerate(copy_data["bins2"].unique()):
+                    bin_mean = copy_data[resp][
+                        (copy_data[pred1] == val1) & (copy_data["bins2"] == x_2)
+                    ].mean()
+                    calc_matrix[x_1, x_2] = bin_mean
+        else:
+            calc_matrix = np.zeros(
+                (len(copy_data[pred1].unique()), len(copy_data[pred2].unique()))
+            )
+            for x_1, val1 in enumerate(copy_data[pred1].unique()):
+                for x_2, val2 in enumerate(copy_data[pred2].unique()):
+                    bin_mean = copy_data[resp][
+                        (copy_data[pred1] == val1) & (copy_data[pred2] == val2)
+                    ].mean()
+                    calc_matrix[x_1, x_2] = bin_mean
+    return calc_matrix
+
+
+def make_clickable(url, name):
+    # This function is from stack overflow link that the professor provided in the slides
+    # https://stackoverflow.com/questions/42263946/how-to-create-a-table-with-clickable-hyperlink-in-pandas-jupyter-notebook
+    return '<a href="{}" rel="noopener noreferrer" target="_blank">{}</a>'.format(
+        url, name
+    )
+
+
+def add_plot_links(table, data, resp):
```

Comment:
This is a really hard way to do this. Try looking for examples with the lambda function. something like this:
df.apply(lambda x: ...). It saves you having to do the looping. You could realistically replace like 50 lines of code in here with like 1 or 2.

Reply:
Actually now that I've looked further down it looks like you have a lambda function which would already be vectorized. It looks like you're then looping through for the whole vector. So if I'm reading this right, you're basically doing a loop within a loop. 

Maybe try breaking this up. I would leave your first function for make_clickable. That's a good idea. You should be able to put the rest of the formatting in the lambda function and then apply it individually to columns. You could pass the column name and the URL column name to the function.